File: common/src/main/java/com/imsproject/common/dataAccess/SQLiteExecutor.java
package com.imsproject.common.dataAccess;

import com.imsproject.common.dataAccess.abstracts.SQLExecutor;
import org.sqlite.SQLiteConfig;

import java.sql.*;
import java.util.Properties;

/**
 * An implementation of {@link SQLExecutor} for SQLite databases
 */
public class SQLiteExecutor implements SQLExecutor {
    private static final String URL_PREFIX = "jdbc:sqlite:";
    private final String URL;
    private final Properties properties;
    private volatile Connection transactionConnection;

    /**
     * @param dbUrl the url of the database to connect to
     */
    public SQLiteExecutor(String dbUrl) {
        URL = URL_PREFIX + dbUrl;
        properties = getProperties();
    }

    private Properties getProperties() {
        SQLiteConfig config = new SQLiteConfig();
        config.enforceForeignKeys(true);
        return config.toProperties();
    }

    @Override
    public void beginTransaction() throws SQLException {
        transactionConnection = DriverManager.getConnection(URL, properties);
        transactionConnection.setAutoCommit(false);
    }

    @Override
    public void commit() throws SQLException {
        if (transactionConnection == null || transactionConnection.isClosed()){
            throw new IllegalStateException("commit() called when not in a transaction");
        }
        try{
            transactionConnection.commit();
        } finally {
            transactionConnection.close();
        }
    }

    @Override
    public void rollback() throws SQLException {

        if (transactionConnection == null || transactionConnection.isClosed()) {
            throw new IllegalStateException("rollback() called when not in a transaction");
        }
        try{
            transactionConnection.rollback();
        } finally {
            transactionConnection.close();
        }
    }

    @Override
    public OfflineResultSet executeRead(String query, Object ... params) throws SQLException {

        if(query == null || query.isBlank()){
            if(inTransaction()){
                rollback();
            }
            throw new SQLException("query is null or empty");
        }

        query = cleanQuery(query);

        if(inTransaction()){
            PreparedStatement statement = transactionConnection.prepareStatement(query);
            prepareStatement(params, statement);
            return new OfflineResultSet(statement.executeQuery());
        } else {
            try (Connection connection = DriverManager.getConnection(URL)) {
                PreparedStatement statement = connection.prepareStatement(query);
                prepareStatement(params, statement);
                ResultSet resultSet = statement.executeQuery();
                return new OfflineResultSet(resultSet);
            }
        }
    }

    @Override
    public int executeWrite(String query, Object ... params) throws SQLException {

        if(query == null || query.isBlank()){
            if(inTransaction()){
                rollback();
            }
            throw new SQLException("query is null or empty");
        }

        query = cleanQuery(query);

        if(inTransaction()){
            try{
                PreparedStatement statement = transactionConnection.prepareStatement(query);
                prepareStatement(params, statement);
                return statement.executeUpdate();
            } catch (SQLException e) {
                rollback();
                throw e;
            }
        } else {
            try (Connection connection = DriverManager.getConnection(URL, properties)) {
                connection.setAutoCommit(false);
                PreparedStatement statement = connection.prepareStatement(query);
                prepareStatement(params, statement);
                try{
                    int rowsChanged = statement.executeUpdate();
                    connection.commit();
                    return rowsChanged;
                }catch (SQLException e){
                    connection.rollback();
                    throw e;
                }
            }
        }
    }

    private boolean inTransaction() throws SQLException {
        return transactionConnection != null && transactionConnection.isClosed() == false;
    }

    private String cleanQuery(String query) {
        query = query.strip();
        query = query.charAt(query.length()-1) == ';' ? query : query + ";";
        return query;
    }

    private void prepareStatement(Object[] params, PreparedStatement s) throws SQLException {
        for(int i = 0; i < params.length; i++){
            s.setObject(i+1, params[i]);
        }
    }
}

File: common/src/main/java/com/imsproject/common/dataAccess/SQLiteExecutor.kt
package com.imsproject.common.dataAccess

import com.imsproject.common.dataAccess.abstracts.SQLExecutor
import org.sqlite.SQLiteConfig
import java.sql.*
import java.util.*
import kotlin.concurrent.Volatile

private const val URL_PREFIX = "jdbc:sqlite:"

/**
 * An implementation of [SQLExecutor] for SQLite databases
 * @param dbUrl the path to the SQLite database file
 */
class SQLiteExecutor (dbUrl: String) : SQLExecutor {
    private val url: String = URL_PREFIX + dbUrl
    private val properties: Properties

    @Volatile
    private var transactionConnection: Connection? = null

    init {
        properties = getProperties()
    }

    @Throws(SQLException::class)
    override fun beginTransaction() {
        val connection = DriverManager.getConnection(url, properties)
        connection.autoCommit = false
        transactionConnection = connection
    }

    @Throws(SQLException::class)
    override fun commit() {
        val transactionConnection = transactionConnection
        check(! (transactionConnection == null || transactionConnection.isClosed)) {
            "commit() called when not in a transaction"
        }
        transactionConnection.use { it.commit() }
    }

    @Throws(SQLException::class)
    override fun rollback() {
        val transactionConnection = transactionConnection
        check(! (transactionConnection == null || transactionConnection.isClosed)) {
            "rollback() called when not in a transaction"
        }
        transactionConnection.use { it.rollback() }
    }

    @Throws(SQLException::class)
    override fun executeRead(query: String, vararg params: Any?): OfflineResultSet {
        if (query.isBlank()) {
            if (inTransaction()) {
                rollback()
            }
            throw SQLException("query is empty")
        }

        val cleanQuery = cleanQuery(query)

        if (inTransaction()) {
            val statement = transactionConnection!!.prepareStatement(cleanQuery)
            prepareStatement(params, statement)
            return OfflineResultSet(statement.executeQuery())
        } else {
            DriverManager.getConnection(url).use { connection ->
                val statement = connection.prepareStatement(cleanQuery)
                prepareStatement(params, statement)
                val resultSet = statement.executeQuery()
                return OfflineResultSet(resultSet)
            }
        }
    }

    @Throws(SQLException::class)
    override fun executeWrite(query: String, vararg params: Any?): Int {
        if (query.isBlank()) {
            if (inTransaction()) {
                rollback()
            }
            throw SQLException("query is empty")
        }

        val cleanQuery = cleanQuery(query)

        if (inTransaction()) {
            try {
                val statement = transactionConnection!!.prepareStatement(cleanQuery)
                prepareStatement(params, statement)
                return statement.executeUpdate()
            } catch (e: SQLException) {
                rollback()
                throw e
            }
        } else {
            DriverManager.getConnection(url, properties).use { connection ->
                connection.autoCommit = false
                val statement = connection.prepareStatement(cleanQuery)
                prepareStatement(params, statement)
                try {
                    val rowsChanged = statement.executeUpdate()
                    connection.commit()
                    return rowsChanged
                } catch (e: SQLException) {
                    connection.rollback()
                    throw e
                }
            }
        }
    }

    /**
     * Executes an insert query by using [java.sql.PreparedStatement]
     * @param query A sql prepared query. Use '?' for parameters
     * @param params The parameters to be used in the query in order
     * @return An [Int] representing the id of the inserted row
     */
    override fun executeInsert(query: String, vararg params: Any?): Int {
        TODO("Not yet implemented")
    }

    /**
     * Executes an update or delete query by using [java.sql.PreparedStatement]
     * @param query A sql prepared query. Use '?' for parameters
     * @param params The parameters to be used in the query in order
     * @return The number of rows affected by the query
     * @throws SQLException if an error occurs while executing the query
     */
    override fun executeUpdateDelete(query: String, vararg params: Any?): Int {
        TODO("Not yet implemented")
    }

    @Throws(SQLException::class)
    private fun inTransaction(): Boolean {
        return transactionConnection != null && transactionConnection!!.isClosed.not()
    }

    private fun cleanQuery(query: String): String {
        var output = query.trim()
        output = if (output[output.length - 1] == ';') output else "$output;"
        return output
    }

    @Throws(SQLException::class)
    private fun prepareStatement(params: Array<out Any?>, s: PreparedStatement) {
        for (i in params.indices) {
            if(params[i] == null){
                s.setNull(i + 1, Types.NULL)
            } else {
                s.setObject(i + 1, params[i])
            }
        }
    }

    private fun getProperties(): Properties {
        val config = SQLiteConfig()
        config.enforceForeignKeys(true)
        return config.toProperties()
    }
}
